File: price_fetcher.py
"""
Price fetching module for MTG card estimator.
Fetches card prices from various sources.
"""
import requests
from typing import Optional, Dict, Any, List


class PriceFetcher:
    """Fetches MTG card prices from online marketplaces."""
    
    def __init__(self):
        """Initialize the price fetcher."""
        self.scryfall_api_base = "https://api.scryfall.com"
        self.session = requests.Session()
        self.session.headers.update({
            'User-Agent': 'MTG-Estimator/1.0'
        })
    
    def get_card_price(self, card_data: Dict[str, Any]) -> Optional[Dict[str, float]]:
        """
        Get price information for a card.
        
        Args:
            card_data: Card data from Scryfall
            
        Returns:
            Dictionary with price information or None
        """
        if not card_data or 'prices' not in card_data:
            return None
        
        prices = card_data['prices']
        result = {}
        
        # Extract available prices
        if prices.get('usd'):
            try:
                result['usd'] = float(prices['usd'])
            except (ValueError, TypeError):
                pass
        if prices.get('usd_foil'):
            try:
                result['usd_foil'] = float(prices['usd_foil'])
            except (ValueError, TypeError):
                pass
        if prices.get('eur'):
            try:
                result['eur'] = float(prices['eur'])
            except (ValueError, TypeError):
                pass
        if prices.get('eur_foil'):
            try:
                result['eur_foil'] = float(prices['eur_foil'])
            except (ValueError, TypeError):
                pass
        
        return result if result else None
    
    def estimate_card_value(self, card_name: str, set_code: Optional[str] = None) -> Optional[float]:
        """
        Estimate the value of a card.
        
        Args:
            card_name: Name of the card
            set_code: Optional set code for specific printing
            
        Returns:
            Estimated USD price or None
        """
        try:
            # Search for the card
            url = f"{self.scryfall_api_base}/cards/named"
            params = {'fuzzy': card_name}
            
            if set_code:
                params['set'] = set_code
            
            response = self.session.get(url, params=params)
            
            if response.status_code == 200:
                card_data = response.json()
                prices = self.get_card_price(card_data)
                
                if prices and 'usd' in prices:
                    return prices['usd']
            
            return None
        except Exception as e:
            print(f"Error estimating card value: {e}")
            return None
    
    def estimate_total_value(self, card_names: List[str]) -> Dict[str, Any]:
        """
        Estimate the total value of multiple cards.
        
        Args:
            card_names: List of card names
            
        Returns:
            Dictionary with individual prices and total
        """
        results = {
            'cards': [],
            'total_usd': 0.0,
            'found': 0,
            'not_found': 0
        }
        
        for card_name in card_names:
            price = self.estimate_card_value(card_name)
            
            if price is not None:
                results['cards'].append({
                    'name': card_name,
                    'price_usd': price
                })
                results['total_usd'] += price
                results['found'] += 1
            else:
                results['cards'].append({
                    'name': card_name,
                    'price_usd': None,
                    'error': 'Price not found'
                })
                results['not_found'] += 1
        
        results['total_usd'] = round(results['total_usd'], 2)
        return results

File: services/price_fetcher.rb
# frozen_string_literal: true

require 'httparty'

# PriceFetcher - Fetches MTG card prices from online marketplaces
class PriceFetcher
  include HTTParty
  base_uri 'https://api.scryfall.com'

  def initialize
    self.class.headers('User-Agent' => 'MTG-Estimator/1.0')
  end

  # Get price information for a card
  # @param card_data [Hash] Card data from Scryfall
  # @return [Hash, nil] Dictionary with price information or nil
  def get_card_price(card_data)
    return nil unless card_data && card_data['prices']

    prices = card_data['prices']
    result = {}

    # Extract available prices
    result['usd'] = prices['usd'].to_f if prices['usd']
    result['usd_foil'] = prices['usd_foil'].to_f if prices['usd_foil']
    result['eur'] = prices['eur'].to_f if prices['eur']
    result['eur_foil'] = prices['eur_foil'].to_f if prices['eur_foil']

    result.empty? ? nil : result
  end

  # Estimate the value of a card
  # @param card_name [String] Name of the card
  # @param set_code [String, nil] Optional set code for specific printing
  # @return [Float, nil] Estimated USD price or nil
  def estimate_card_value(card_name, set_code = nil)
    params = { fuzzy: card_name }
    params[:set] = set_code if set_code

    response = self.class.get('/cards/named', query: params)

    if response.success?
      card_data = response.parsed_response
      prices = get_card_price(card_data)
      
      return prices['usd'] if prices && prices['usd']
    end

    nil
  rescue StandardError => e
    puts "Error estimating card value: #{e.message}"
    nil
  end

  # Estimate the total value of multiple cards
  # @param card_names [Array<String>] List of card names
  # @return [Hash] Dictionary with individual prices and total
  def estimate_total_value(card_names)
    results = {
      'cards' => [],
      'total_usd' => 0.0,
      'found' => 0,
      'not_found' => 0
    }

    card_names.each do |card_name|
      price = estimate_card_value(card_name)

      if price
        results['cards'] << {
          'name' => card_name,
          'price_usd' => price
        }
        results['total_usd'] += price
        results['found'] += 1
      else
        results['cards'] << {
          'name' => card_name,
          'price_usd' => nil,
          'error' => 'Price not found'
        }
        results['not_found'] += 1
      end
    end

    results['total_usd'] = results['total_usd'].round(2)
    results
  end
end
